docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,99 @@
+
+Here is the proposed set of `ConnectionManager` states and how they map to the `Controller` states. Please note that some of these states may go away once we are further along the implementation (e.g. we may not need `ConnectionStateSilentSwitching` anymore):
+
+```c++
+  enum ConnectionState {
+    ConnectionStateOff,  // Corresponds to controller states: StateInitializing,
+                         // StateOff
+    ConnectionStateInternetProbe,      // Corresponds to controller states:
+                                       // StateConnecting
+    ConnectionStateServerProbe,        // Corresponds to controller states:
+                                       // StateConnecting
+    ConnectionStateFirewall,           // Corresponds to controller states:
+                                       // StateConnecting
+    ConnectionStateCaptivePortal,      // Corresponds to controller states:
+                                       // StateConnecting
+    ConnectionStateCheckSubscription,  // Corresponds to controller states:
+                                       // StateCheckSubscription
+    ConnectionStateOn,  // Corresponds to controller state: StateOn
+
+    // Mirrored from Controller states
+    ConnectionStateSilentSwitching,  // Corresponds to StateSilentSwitching
+    ConnectionStateSwitching,        // Corresponds to StateSwitching
+    ConnectionStateConfirming,       // Corresponds to controller states:
+                                     // StateConfirming
+    ConnectionStateDisconnecting     // Corresponds to controller state:
+                                     // StateDisconnecting
+  };
+```
+
+If at any point any of the checks in `ConnectionManager::ConnectionStates` fails then we will emit a signal to the frontend explaining what exactly the issue is. For example, if we are in `ConnectionManager::ConnectionStateInternetProbe` and the internet probe fails, we will notify the frontend via `emit connectionStateInternetProbeFailed()` signal. This allows the UI to notify the user that the VPN activation failed due to their network connectivity issue and may offer troubleshooting advice such as checking that their wifi/cellular is enabled. It is worth noting that in any given circumstance, we will not deactivate the VPN on our end regardless of the failure. 
+
+This change also means that prior to the activation of the VPN, there are multiple checks and probes that need to succeed; this will take some time (hopefully no more than a couple of seconds), and as long as we communicate to the user via front end prompts about what is happening, this is acceptable.
```

Comment:
If the hope that this doesn't take a long time is informing the proposed solution here, we should have a plan to transform that hypothesis into a fact. Let's add a section here that specs out some telemetry that will be added to monitor that this assumption is correct.